docs/markdown/Build-options.md

```diff
@@ -60,6 +61,36 @@ default.
 
 This type is available since version 0.44.0
 
+### Features
+
+A `feature` option has three states: `enabled`, `disabled` or
+`automatic`. It is intended to be passed as value for the `required` keyword
+argument of most functions. Currently supported in
+[`dependency()`](Reference-manual.md#dependency),
+[`find_library()`](Reference-manual.md#compiler-object),
+[`find_program()`](Reference-manual.md#find_program) and
+[`add_languages()`](Reference-manual.md#add_languages) functions.
+
+- `enabled` is the same as passing `required: true`.
+- `automatic` is the same as passing `required: false`.
+- `disabled` cause it to always claim the dependency couldn't be found.
```

Comment:
There are two different cases here: it can return either a not-found dependency (even if the dependency could be found) or a disabler. Should both of these be possible to do?

Reply:
That's PR #3386.

Reply:
Hmmmmm. I'm on the fence on whether they should be separate kwargs or should there be only one...

Reply:
They would be mutually exclusive, that's a bit weird too.

Reply:
> Hmmmmm. I'm on the fence on whether they should be separate kwargs or should there be only one...

Are you speaking about the `use_disabler` or the `required` kwarg?

If what you mean is keeping `required` kwarg as boolean, and add an extra kwarg for feature option (that's what you did in the dolphin PR) I think it's a bit weird too because those 2 kwarg would be mutually exclusive.

Reply:
What I mean is that an alternative approach to the current setup with a separate disabler kwarg would be to only have this kwarg but the option would have four different choices:

 - `required`
 - `auto`
 - `return_not_found_without_looking`
 - `return_disabler`

Reply:
Ah ok, that was my very first idea when I started working on this but I changed my mind on this. The reason is that it's not up to the user to decide the mechanism to use but up to the project developer.

For example if the developer wants to rely on the disabler mechanism he would code like this and expect to have either no app built or app with dep:
```meson
dep = dependency('foo', required: get_option('foo'))
app = executable('app', dependencies : dep)
```
But then if user does `-Dfoo=return_not_found_without_looking` it will fail to build app because app cannot be built without dep.

So if we have 4 states it means the project must support both code paths, which completely defeat the point of having disabler to simplify the build system.

```meson
dep = dependency('foo', required: get_option('foo'))
if dep.found()
  app = executable('app', dependencies : dep)
endif
```

Reply:
Note that one better possibility here is to have 3 states enabled/disabled/auto, but `get_option('a-disabled-feature')` could actually return a `disabler` object. That would mean no change is needed to dependency() because it would receive either true/false boolean or `disabler` object.

I decided against that idea too for 2 reasons:
- Using disabler is nice but often require to refactor a lot your meson.build. Even if the result is probably often cleaner, that's still more work to get feature option adopted.
- the return value of get_option() becomes much harder to use if it's not passed directly to `required` kwarg. You cannot do stuff like:
```meson
if get_option('a-disabled-option').disabled()
....
endif
```
because you whole if block gets ignored.

Reply:
Converting a "not found" into a disabler (I'm assuming that you forgot to add that kwarg to the snippet above) does have use on its own. In fact, I'd wager that it would become fairly popular for some dependencies.

If we wanted to replicate that with the option it would need to have five states:

 - `auto`
 - `required`
 - `return_not_found_without_looking`
 - `return_disabler`
 - `return_disabler_if_not_found`

Which gets overly complicated. The two kwarg approach seems like the correct thing to do here.

Reply:
> d = dependency('foobar', use_disabler : true)

Hm, I'm not sure about returning disabler for a not-found required dep. It means you can do something if it's not-found... then you should have `required:false`...

Reply:
Those two are unrelated. You are right that it should probably have `required : false`.

The 5th state is what we would need to have if we don't have `use_disabler` kwarg. But we should, since it simplifies things a lot.